include/grpc/support/port_platform.h

```diff
@@ -53,6 +53,57 @@
 #define WIN32_LEAN_AND_MEAN
 #endif /* WIN32_LEAN_AND_MEAN */
 
+// GPRC_DLL
+// inspired by
+// https://github.com/abseil/abseil-cpp/blob/20220623.1/absl/base/config.h#L730-L747
+//
+// When building gRPC as a DLL, this macro expands to `__declspec(dllexport)`
+// so we can annotate symbols appropriately as being exported. When used in
+// headers consuming a DLL, this macro expands to `__declspec(dllimport)` so
+// that consumers know the symbol is defined inside the DLL. In all other cases,
+// the macro expands to nothing.
+//
+// Warning: shared library support for Windows (i.e. producing DLL plus import
+//   library instead of a static library) is experimental. Some symbols that can
+//   be linked using the static library may not be available when using the
+//   dynamically linked library.
+//
+// Note: GRPC_DLL_EXPORTS is set in CMakeLists.txt when building shared
+// grpc{,_unsecure}
+//       GRPC_DLL_IMPORTS is set by us as part of the interface for consumers of
+//       the DLL
+#if !defined(GRPC_DLL)
```

Comment:
note that the GRPC_DLL GRPCXX_DLL and GPR_DLL defines depend on the assumption that the only 3  "public" libraries known to us are gpr, grpc and grpc++. In the future, we are likely planning to make the layout of libraries in cmake more akin to what's in grpc's bazel BUILD (that has much more granular targets) at which point it won't be feasible to hand list all the known public libraries here?

I'll leave the decision whether the defines make sense as-is to @veblush  but I wanted to point out this potential problem.

Reply:
at the same time, I realize that because of the import / export semantics in windows DLLs, there probably needs to be a separate define for each library, so this is tricky to resolve.
What does absl do here (since AFAIK absl consists of many shared libraries)?

Reply:
> at the same time, I realize that because of the import / export semantics in windows DLLs, there probably needs to be a separate define for each library, so this is tricky to resolve. What does absl do here (since AFAIK absl consists of many shared libraries)?

Abseil uses a single pair of macros for import/export determination ABSL_BUILD_DLL & ABSL_CONSUME_DLL

The abseil shared build emits a single DLL: abseil_dll.dll with an import library of abseil_dll.lib. It also emits other more specific import libraries, but I believe that these still reference the same DLL

Reply:
> I realize that because of the import / export semantics in windows DLLs, there probably needs to be a separate define for each library, so this is tricky to resolve.
> What does absl do here (since AFAIK absl consists of many shared libraries)?

I believe the pain with having a separate symbol per library is why abseil essentially creates one big DLL (with consequently only one macro switching between dllimport & dllexport), rather than handling this on a per-library level.

grpc's current situation with 3 libs is perhaps still acceptable for a manual setup, but if there are going to be more, then the shared setup on windows will likely have to bunch them together.

Note that this is a "lesser-of-two-evils" situation, because the build divergence between `unix || static` and `win && shared` causes quite a bit of hassle, both for the build itself, but also for the pkgconfig & CMake metadata.

CC @derekmauro

Reply:
I think this is acceptable as we don't want to have many DLLs even after refactoring the cmake build to have 1:1 corresponding targets to bazel's.